lint ImproperCTypes: refactor linting architecture (part 2)

niacdoial opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/146273)*
<!-- homu-ignore:end -->

This is the second PR in an effort to split https://github.com/rust-lang/rust/pull/134697 (refactor plus overhaul of the ImproperCTypes family of lints) into individually-mergeable parts.

Contains the changes of the first PR, and splits the core type checking function into several bits, each focused on a specific aspect of FFI-safety.
Some logic which was outside of said core function was also moved into the new functions.

Superset of: rust-lang/rust#146271

tgross35 reviewed (changes requested):
I think a lot of this makes sense, but aren't there behavior changes here? It looks like tuples and arrays may be treated slightly differently.

Which is probably fine, that would ideally just be split from the refactoring and come with test updates.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146273/0d0f4eac8b98133e5da6d3604d86a8f3b5a67844..1dbb0e2a79586d1ebab2d99c0a70b1b682564771)*

niacdoial commented:
> I think a lot of this makes sense, but aren't there behavior changes here? It looks like tuples and arrays may be treated slightly differently.

> Which is probably fine, that would ideally just be split from the refactoring and come with test updates.

I moved the actual change in behaviour in a later commit.
The rest of the changes here are just an exercise in moving more of the type-checking logic into the `visit_*` methods.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17519237695/job/49760866076) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/49760866076) [(plain)](https://github.com/rust-lang/rust/commit/f54061c4ea04022d134b2e2c7ddca1b6a6000894/checks/49760866076/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_passes v0.0.0 (/checkout/compiler/rustc_passes)
error[E0061]: this method takes 3 arguments but 1 argument was supplied
   --> compiler/rustc_lint/src/types/improper_ctypes.rs:666:40
    |
666 |             ty::Pat(pat_ty, _) => self.visit_type(pat_ty),
    |                                        ^^^^^^^^^^-------- two arguments of type `VisitorState` and `std::option::Option<rustc_middle::ty::Ty<'_>>` are missing
    |
note: method defined here
   --> compiler/rustc_lint/src/types/improper_ctypes.rs:615:8
    |
615 |     fn visit_type(
    |        ^^^^^^^^^^
616 |         &mut self,
617 |         state: VisitorState,
    |         -------------------
618 |         outer_ty: Option<Ty<'tcx>>,
    |         --------------------------
help: provide the arguments
    |
666 |             ty::Pat(pat_ty, _) => self.visit_type(/* VisitorState */, /* std::option::Option<rustc_middle::ty::Ty<'_>> */, pat_ty),
    |                                                   ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

For more information about this error, try `rustc --explain E0061`.
[RUSTC-TIMING] rustc_lint test:false 3.103
error: could not compile `rustc_lint` (lib) due to 1 previous error
```

</details>

niacdoial commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -446,29 +632,18 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
+                        //   - not using an indirection would cause a compile error (this lint *currently* seems to not get triggered on such non-compiling code)
+                        //   - and using one would cause the lint to catch on the indirection before reaching its pointee
+                        // but function *pointers* don't seem to have the same no-unsized-parameters requirement to compile
                         if let Some(sym::cstring_type | sym::cstr_type) =
```
it's not describing the behaviour, it's just justifying that "yes, we can also encounter sym::cstr_type in this piece of code.
(I'm removing the comment for now, it's not relevant until indirections get treated differently)

niacdoial commented:
> `Array` went from just checking the type to checking whether or not it is in a function. 

That's a bit of logic that was moved from `check_type` to `visit_type`

niacdoial committed:
ImproperCTypes: merge outer_ty information into VisitorState

Another user-transparent change, unifying outer-type information and the
existing VisitorState flags.

niacdoial commented:
ah, I knew I was missing something

tgross35 commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -274,9 +298,40 @@ bitflags! {
+    /// Flags describing both the immediate and overall context in which the current mir::Ty is
+    persistent_flags: PersistentStateFlags,
+    /// Flags describing both the immediate and overall context in which the current mir::Ty is
+    ephemeral_flags: EphemeralStateFlags,
```
The two doc comments here are the same

tgross35 commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -291,57 +346,133 @@ impl VisitorState {
+        }
+    }
+    fn get_next_in_fnptr<'tcx>(&self, current_ty: Ty<'tcx>, is_ret: bool) -> Self {
+        debug_assert!(matches!(current_ty.kind(), ty::FnPtr(..)));
```
These can be real asserts, they should be easy for the compiler to remove (since the enum variant is always checked before calling).

It would be good to mention the difference between these two functions in the docs for `get_next` as well.

tgross35 commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -291,57 +346,133 @@ impl VisitorState {
+            ephemeral_flags: EphemeralStateFlags::from_outer_ty(current_ty),
+        }
+    }
+    fn get_next_in_fnptr<'tcx>(&self, current_ty: Ty<'tcx>, is_ret: bool) -> Self {
```
Optional nit: adding an enum rather than using a bool may come in handy in a few places

```rust
enum FnPos {
    Arg,
    Ret,
}
```

tgross35 commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -364,36 +495,94 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
+                // FIXME(ctypes): this logic is broken, but it still fits the current tests:
+                // - for some reason `Box<_>`es in `extern "ABI" {}` blocks
+                //   (including within FnPtr:s)
+                //   are not treated as pointers but as unsafe structs
```
Docs nit

```suggestion
                // - for some reason `Box<_>`es in `extern "ABI" {}` blocks
                //   (including within FnPtr:s)
                //   are not treated as pointers but as FFI-unsafe structs
```

tgross35 reviewed (changes requested):
Sorry for the delay here, I've been behind for al little while. Should be catching up now, though

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146273/59043567a5cf12800e1457c36ad6a6b0fa02c6b6..359cb792b0c82fc6ebace123e0e95d78d75d1c06)*

niacdoial commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -274,9 +298,40 @@ bitflags! {
+    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
+    struct EphemeralStateFlags: u8 {
+        /// To annotate pointees (through Ref,RawPtr,Box).
+        const IN_PTR = 0b000001;
```
it's building on the mental model of "entering" and "exiting" types as they are visited: the flag `IN_ADT` tells that the currently checked at this depth is a field to the ADT that's "one layer outside".
same with `IN_PTR`, saying that we are dealing with the pointee of what's "one layer outside".
...though now that you point it out, the flags' names make them sound like they would also be applied recursively when checking "further in", which is not the case.

niacdoial commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -291,57 +346,133 @@ impl VisitorState {
+
+impl EphemeralStateFlags {
+    /// modify self to change the ephemeral part of the flags
+    fn from_outer_ty<'tcx>(ty: Ty<'tcx>) -> Self {
```
yep, tastes like docs I forgot to update as I was working on that abstraction.

niacdoial commented:
no worries for the delay!
I even had the time to think about this whole thing in the background, and find a couple of points I'm no longer happy with, in commits that are only in the original "take 2" PR.  (:
I'll try adding a more tests to get full coverage of the linting code, too

tgross35 commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -291,57 +346,133 @@ impl VisitorState {
+    }
+
+    /// Get the proper visitor state for a static variable's type
+    #[inline]
```
Nit: no need to `#[inline]`, that will happen automatically on crate-private functions

tgross35 commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -291,57 +346,133 @@ impl VisitorState {
+            && self.is_in_function_return()
+    }
+
+    /// Whether the type is directly used in a function.
```
```suggestion
    /// Whether the type is a function argument or return type, without indirection.
```

tgross35 commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -429,6 +619,111 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
+                help: if def.is_struct() {
+                    Some(fluent::lint_improper_ctypes_struct_layout_help)
+                } else {
+                    // FIXME(ctypes): confirm that this makes sense for unions once #60405 / RFC2645 stabilises
```
```suggestion
                    // FIXME(#60405): confirm that this makes sense for unions once #60405 / RFC2645 stabilises
```

Since this will be fixed by something other than the ctypes refactoring

tgross35 commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -274,9 +298,40 @@ bitflags! {
+    persistent_flags: PersistentStateFlags,
+    /// Flags describing both the immediate and overall context in which the current mir::Ty is
+    ephemeral_flags: EphemeralStateFlags,
+}
```
The usability argument is fair. However, my concern is that we could accidentally wind up with a meaningless state like `MEMORY_INLINED | IN_PTR`, and this becomes a lot easier to understand if we don't need to think about that.

I'm realizing that a lot of these flags aren't yet consumed: many are set but only `NO_OUTER_TY` `NOOUT_FNPTR` are checked against. So I think we can also start a bit smaller here and build up in the other PRs. How about replacing `EphemeralStateFlags` with:

```rust
/// State representing the parent of immediate type, if any.
enum Indirection {
    /// There is no parent type, or the parent type is not relevant.
    // FIXME(ctypes): there are actually more states we need to worry about
    // here, but these are in the process of being added.
    None,
    /// The type is an argument or return type in a function pointer.
    // FIXME(ctypes): this is only used for compatibility, and will be removed with a behavior fix.
    FnPtr,
}
```

Then we can fill in the rest as needed, we'll be able to add a `Pointer { /* pointer-specific flags */ }` variant later on.

tgross35 commented:
I just stumbled upon https://github.com/rust-lang/rust/issues/66220, that thread is probably worth a read. I think I understand the difference between the two lints better now:

- `improper_ctypes` is the original and is more strict, it checks the types used even through indirection
- `improper_ctypes_definitions` allows non-FFI-safe types if protected via indirection

Which I think explains some of the current behavior a bit better. Eventually I wonder if it might make sense to eventually add a variant to the result type, so

```rust
enum FfiResult<'tcx> {
    FfiSafe,
    FfiPhantom(Ty<'tcx>),
    /// Contains FFI-unsafe types, made okay to use in a signature by a pointer type.
    SafeViaIndirection { ty: Ty<'tcx>, reason: DiagMessage, help: Option<DiagMessage> },
    /// FFI-unsafe
    FfiUnsafe { ty: Ty<'tcx>, reason: DiagMessage, help: Option<DiagMessage> },
}
```

Sort of as part of what is suggested in https://github.com/rust-lang/rust/issues/72774. But that would be something for later.

niacdoial commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -259,9 +271,21 @@ enum FfiResult<'tcx> {
+    /// "Permanent state" flags for VisitorState (kept when visiting new mir::Ty)
     #[derive(Clone, Copy, Debug, PartialEq, Eq)]
-    struct VisitorState: u8 {
+    struct PersistentStateFlags: u16 {
```
hm... how about `RootUseFlags`?

niacdoial replied:
it would be short for "how the root type is used by the FFI boundary"
and... do we have a separate DefId for every function argument/return / static variable / etc?

niacdoial commented:
ok! all is split, with the two commits "before" this one being https://github.com/rust-lang/rust/pull/155358 and https://github.com/rust-lang/rust/pull/155359
sorry it took some time, I ended up being far more busy than I thought I would

petrochenkov commented:
Blocked on https://github.com/rust-lang/rust/pull/155359.
@rustbot blocked

petrochenkov commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -302,7 +314,45 @@ bitflags! {
+            | ty::Tuple(..)
+            | ty::Array(..)
+            | ty::Slice(_) => Self::UnusedVariant,
+            k @ _ => bug!("Unexpected outer type {:?} of kind {:?}", ty, k),
```
```suggestion
            _ => bug!("unexpected outer type {ty:?}),
```
Debug printing for `Ty` already redirects to `TyKind` printing.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146273/d2f88d4b65f64f61293b739bc0c7f97b3a7805a8..ac62a57b383d0fe6ed020b8e9911b365e8199095)*

petrochenkov commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -302,7 +314,45 @@ bitflags! {
+
+impl OuterTyKind {
+    /// Computes the relationship by providing the containing mir::Ty itself
+    fn from_outer_ty<'tcx>(ty: Ty<'tcx>) -> Self {
```
```suggestion
    fn from_ty<'tcx>(ty: Ty<'tcx>) -> Self {
```
Nit: already have "outer" in `OuterTyKind`.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146273/d2f88d4b65f64f61293b739bc0c7f97b3a7805a8..ac62a57b383d0fe6ed020b8e9911b365e8199095)*

petrochenkov commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -302,7 +314,45 @@ bitflags! {
+    /// but is needed because we don't change the lint's behavior yet
+    NoneThroughFnPtr,
+    /// Placeholder for properties that will be used eventually
+    UnusedVariant,
```
```suggestion
    Other,
```
It's sort of used, because it indicates that the parent type is one of the *allowed* ones (and we panic on non-allowed). The comment can be also updated to "other allowed outer type".

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146273/d2f88d4b65f64f61293b739bc0c7f97b3a7805a8..ac62a57b383d0fe6ed020b8e9911b365e8199095)*

petrochenkov commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -317,86 +367,100 @@ impl VisitorState {
+impl VisitorState {
+    /// From an existing state, compute the state of any subtype of the current type.
+    /// (General case. For the case where the current type is a function pointer, see `get_next_in_fnptr`.)
+    fn get_next<'tcx>(&self, current_ty: Ty<'tcx>) -> Self {
```
```suggestion
    fn next(&self, current_ty: Ty<'_>) -> Self {
```

- `get_` is usually omitted in typical naming conventions because it's kind of useless.
- some of the lifetimes here and below are *probably* unnecessary.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146273/d2f88d4b65f64f61293b739bc0c7f97b3a7805a8..ac62a57b383d0fe6ed020b8e9911b365e8199095)*

petrochenkov commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -317,86 +367,100 @@ impl VisitorState {
+    /// (General case. For the case where the current type is a function pointer, see `get_next_in_fnptr`.)
+    fn get_next<'tcx>(&self, current_ty: Ty<'tcx>) -> Self {
+        assert!(!matches!(current_ty.kind(), ty::FnPtr(..)));
+        Self {
```
```suggestion
        VisitorState {
```
Could you avoid `Self` in non-generic code? At least in constructors like this.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146273/d2f88d4b65f64f61293b739bc0c7f97b3a7805a8..ac62a57b383d0fe6ed020b8e9911b365e8199095)*

niacdoial replied:
OK! I'm also changing the type annotation for OuterTyKind::from_ty's return and the RootUseFlags consts. Is this something you want?
edit: `x fmt` undoes that so...

petrochenkov commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -317,86 +367,100 @@ impl VisitorState {
+            root_use_flags: self.root_use_flags,
+            outer_ty_kind: OuterTyKind::from_outer_ty(current_ty),
         }
     }
```
Mega-nit: space between functions.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146273/d2f88d4b65f64f61293b739bc0c7f97b3a7805a8..ac62a57b383d0fe6ed020b8e9911b365e8199095)*

petrochenkov commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -317,86 +367,100 @@ impl VisitorState {
+    /// Generate the state for an "outermost" type that needs to be checked
+    fn entry_point(root_use_flags: RootUseFlags) -> Self {
+        Self { root_use_flags, outer_ty_kind: OuterTyKind::None }
+    }
```
I'd inline and remove this function, it's trivial and is only used twice.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146273/d2f88d4b65f64f61293b739bc0c7f97b3a7805a8..ac62a57b383d0fe6ed020b8e9911b365e8199095)*

petrochenkov commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -317,86 +367,100 @@ impl VisitorState {
+    }
+
+    /// Get the proper visitor state for a static variable's type
+    fn static_var() -> Self {
```
```suggestion
    fn static_entry_point() -> Self {
```
Nit: I don't think these are called "variables" often in Rust.
Also for consistency with `fn_entry_point`.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146273/d2f88d4b65f64f61293b739bc0c7f97b3a7805a8..ac62a57b383d0fe6ed020b8e9911b365e8199095)*

petrochenkov commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -317,86 +367,100 @@ impl VisitorState {
+    }
+
+    /// Whether the type itself is the type of a function argument or return type.
+    fn is_direct_in_function(&self) -> bool {
```
I'd rather not introduce simple functions that are used once, this one and `is_direct_function_return` can be inlined and removed.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146273/d2f88d4b65f64f61293b739bc0c7f97b3a7805a8..ac62a57b383d0fe6ed020b8e9911b365e8199095)*

petrochenkov commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -317,86 +367,100 @@ impl VisitorState {
+    }
+
+    /// Get the proper visitor state for a given function's arguments or return type.
+    fn entry_point_from_fnmode(fn_mode: CItemKind, fn_pos: FnPos) -> Self {
```
```suggestion
    fn fn_entry_point(fn_mode: CItemKind, fn_pos: FnPos) -> Self {
```
For consistency with `static_entry_point`.
The `_from_fnmode` suffix would make sense if there were other simpler `fn_entry`, not from fn mode.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146273/d2f88d4b65f64f61293b739bc0c7f97b3a7805a8..ac62a57b383d0fe6ed020b8e9911b365e8199095)*

petrochenkov commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -364,36 +387,88 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
+            },
+            _ => bug!("visit_scalar is to be called with scalar (char, int, float, bool) types"),
+        }
+    }
```
No longer relevant.

petrochenkov commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -41,3 +183,946 @@ fn variant_has_complex_ctor(variant: &ty::VariantDef) -> bool {
+                        // most of those are not even reachable,
+                        // but let's not worry about checking that here
+                        _ => false,
+                    }
```
No longer relevant.

petrochenkov commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -274,9 +298,40 @@ bitflags! {
+        const IN_ADT = 0b010000;
+        /// To show that there is no outer type, the current type is directly used by a `static`
+        /// variable or a function/FnPtr
+        const NO_OUTER_TY = 0b100000;
```
No longer relevant.

niacdoial committed:
ImproperCTypes: Use more `Unnormalized` type wrappers

In order to follow along with the efforts to properly distinguish
already-normalised and unnormalized types, we separate the internal
interfaces of this lint that rely on normalized types from those that
do not. We do that by adding the `Unnormalized` wrapper to some
interfaces.

niacdoial commented:
added a new commit to better make use of the `Unnormalized<>` type wrapper that appeared during one of the rebases for this PR and its sub-PRs

niacdoial commented:
wait, have I not...?
@rustbot review

petrochenkov commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -286,8 +295,10 @@ enum IndirectionKind {
+    /// (These are the permanent part of the state, kept when visiting new mir::Ty.)
     #[derive(Clone, Copy, Debug, PartialEq, Eq)]
-    struct VisitorState: u8 {
+    struct RootUseFlags: u16 {
```
```suggestion
    struct RootUseFlags: u8 {
```
Extending to 16 bit is not (yet?) necessary.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146273/20de910db49d3476ccf49ea79a4b22e2b5dface0..03002778f4f0d6f2a257d37f0f011f7df43705da)*

petrochenkov commented on compiler/rustc_lint/src/types/improper_ctypes.rs:
```diff
@@ -286,8 +295,10 @@ enum IndirectionKind {
 
 bitflags! {
+    /// VisitorState flags that are linked with the root type's use.
+    /// (These are the permanent part of the state, kept when visiting new mir::Ty.)
```
```suggestion
    /// (These are the permanent part of the state, kept when visiting new `Ty`.)
```
I've just noticed, what is mir::Ty?
In this PR it seems to refer to the usual `Ty`, which is not "mir".

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146273/20de910db49d3476ccf49ea79a4b22e2b5dface0..03002778f4f0d6f2a257d37f0f011f7df43705da)*

niacdoial replied:
ah, my bad.
It was to differentiate from `hir::Ty` (`rustc_hir::ty::Ty`) used elsewhere in the file, and I kinda assumed this meant the usual Ty (`rustc_middle::ty::Ty`) is MIR.

Should I still make it obvious this isn't the same as `hir::Ty`?

petrochenkov replied:
If you need to disambiguate, then ty::Ty or hir::Ty can be used (whatever requires less annotations).

petrochenkov commented:
r=me after addressing the remaining nits.
@rustbot author

niacdoial commented:
@bors r=petrochenkov

petrochenkov commented:
@bors r+

JonathanBrouwer commented:
@bors yield
Yielding to enclosing rollup

JonathanBrouwer commented:
@bors yield
Yielding to enclosing rollup

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-21-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/25476451519/job/74750942750) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/74750942750) [(plain)](https://github.com/rust-lang/rust/commit/74c42e9f33148e9ae927143cff47f5c33f634fb8/checks/74750942750/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test expr_store::scope::tests::test_shadow_variable ... ok
test expr_store::scope::tests::test_method_call_scope ... ok
test expr_store::tests::body::array_element_cfg ... ok
test expr_store::tests::body::block::inner_item_smoke ... ok
error: test failed, to rerun pass `-p hir-def --lib`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/hir_def-c439368f2b64a049 '--skip=tests::smoke_test_real_sysroot_cargo' --skip=check_code_formatting -Z unstable-options --format json` (signal: 11, SIGSEGV: invalid memory reference)
Bootstrap failed while executing `--stage 2 test --skip tests --skip coverage-map --skip coverage-run --skip library --skip tidyselftest`
Build completed unsuccessfully in 0:49:27
  local time: Thu May  7 05:37:38 UTC 2026
  network time: Thu, 07 May 2026 05:37:38 GMT
##[error]Process completed with exit code 1.
```

</details>

JonathanBrouwer commented:
@bors retry
